dissect/target/plugins/os/windows/dpapi/credhist.py

```diff
@@ -0,0 +1,134 @@
+from dataclasses import dataclass, field
+from io import BytesIO
+from pathlib import Path
+from typing import Iterator, Optional
+from uuid import UUID
+
+from dissect.cstruct import cstruct
+from dissect.util.sid import read_sid
+
+from dissect.target.exceptions import UnsupportedPluginError
+from dissect.target.plugin import InternalPlugin
+from dissect.target.plugins.os.windows.dpapi.crypto import (
+    CipherAlgorithm,
+    HashAlgorithm,
+    derive_password_hash,
+)
+from dissect.target.target import Target
+
+credhist_def = """
+struct entry {
+    DWORD   dwVersion;
+    CHAR    guidLink[16];
+    DWORD   dwNextLinkSize;
+    DWORD   dwCredLinkType;
+    DWORD   algHash;                    // ALG_ID
+    DWORD   dwPbkdf2IterationCount;
+    DWORD   dwSidSize;
+    DWORD   algCrypt;                   // ALG_ID
+    DWORD   dwShaHashSize;
+    DWORD   dwNtHashSize;
+    CHAR    pSalt[16];
+    CHAR    pSid[dwSidSize];
+    CHAR    encrypted[(dwShaHashSize + dwNtHashSize) + (dwShaHashSize + dwNtHashSize) % 16];
+};
+"""
+
+c_credhist = cstruct()
+c_credhist.load(credhist_def)
+
+
+@dataclass
+class CredHistEntry:
+    version: int
+    guid: str
+    user_sid: str
+    hash_alg: any = field(repr=False)
+    cipher_alg: any = field(repr=False)
+    hash_sha: Optional[str]
+    hash_nt: Optional[str]
+    raw: c_credhist.entry = field(repr=False)
+
+    def decrypt(self, password_hash: bytes) -> None:
+        data = self.cipher_alg.decrypt_with_hmac(
+            data=self.raw.encrypted,
+            key=derive_password_hash(password_hash, self.user_sid),
+            iv=self.raw.pSalt,
+            hash_algorithm=self.hash_alg,
+            rounds=self.raw.dwPbkdf2IterationCount,
+        )
+
+        sha_size = self.raw.dwShaHashSize
+        nt_size = self.raw.dwNtHashSize
+
+        self.hash_sha = data[:sha_size].hex()
+        self.hash_nt = data[sha_size : sha_size + nt_size].rstrip(b"\x00").hex()
+
+
+class CredHistFile:
```

Comment:
I'm not feeling the current setup with this class and the dataclass just yet, but I also couldn't think of a better way right now. Just noting my thoughts in case maybe someone else has a better idea.